stat_summary: support service metrics using `authority` label

Pothulapati opened the pull request:
Currently, `viz stat` on services is pretty restricted because of
it not being a podowner resource. This PR fixes that by making
it use the `direction="outbound", authroty="svc"` while querying
the prometheus metrics. This means that for services, we can
generate metrics from the *meshed* clients side.

`StatsSummary` metrics on a service are further divided into
two kinds

### Service has no `ServiceProfiles.dstOverrides` 

In this case, We just return the metrics by
querying for `direction="outbound", authroty="svc"`, along
with any `--from` resources specified as client query labels.

We also gate this path, to fail for requests that have `--from`
as a service or for `svc/* --to xyz`, as they are invalid i.e 
we can't render metrics with service as the client.

### Service has `ServiceProfiles.dstOverrides` 

Here, We follow a similar path of `TrafficSplit`
except that we use a `ServiceProfile` resource
object instead.

_The TrafficSplit path will be removed or merged into the 
`Service` path in a separate PR for simplification,_


## Testing

### Apply Traffic Splitting through `ServiceProfiles`

```bash
on ⛵ kind-kind  linkerd2 on 🌱 taru [📦++1🤷‍] via 🐼 v1.16.5 took 1m11s
➜ k create ns linkerd-trafficsplit-test-sp                                                                                                                                ~/work/linkerd2
namespace/linkerd-trafficsplit-test-sp created

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦++1🤷‍] via 🐼 v1.16.5
➜ ./bin/linkerd inject ./test/integration/trafficsplit/testdata/application.yaml | k -n linkerd-trafficsplit-test-sp apply -f -                                           ~/work/linkerd2

document missing "kind" field, skipped
deployment "backend" injected
service "backend-svc" skipped
deployment "failing" injected
service "failing-svc" skipped
deployment "slow-cooker" injected
service "slow-cooker" skipped

deployment.apps/backend created
service/backend-svc created
deployment.apps/failing created
service/failing-svc created
deployment.apps/slow-cooker created
service/slow-cooker created

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦++1🤷‍] via 🐼 v1.16.5
➜ k apply -f ./test/integration/trafficsplit/testdata/sp/updated-traffic-split-leaf-weights.yaml -n linkerd-trafficsplit-test-sp                                          ~/work/linkerd2
serviceprofile.linkerd.io/backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local created

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦++1🤷‍] via 🐼 v1.16.5
➜ k describe sp -n linkerd-trafficsplit-test-sp                                                                                                                           ~/work/linkerd2
Name:         backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local
Namespace:    linkerd-trafficsplit-test-sp
Labels:       <none>
Annotations:  <none>
API Version:  linkerd.io/v1alpha2
Kind:         ServiceProfile
Metadata:
  Creation Timestamp:  2021-07-01T11:05:06Z
  Generation:          1
  Managed Fields:
    API Version:  linkerd.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:dstOverrides:
    Manager:         kubectl-client-side-apply
    Operation:       Update
    Time:            2021-07-01T11:05:06Z
  Resource Version:  1398
  UID:               fce0a250-1396-4a14-9729-e19030048c7a
Spec:
  Dst Overrides:
    Authority:  backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local
    Weight:     500m
    Authority:  failing-svc.linkerd-trafficsplit-test-sp.svc.cluster.local:8081
    Weight:     500m
Events:         <none>
```

### CLI Output

```bash
on ⛵ kind-kind  linkerd2 on 🌱 main [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp                                                                                                     ~/work/linkerd2
NAME                                                         APEX                                                         LEAF          WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc     500m   100.00%   0.9rps           1ms           2ms           2ms
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   failing-svc     500m     0.00%   1.1rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 main [📦📝🤷‍] via 🐼 v1.16.6 via  took 2s
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp --from deploy/slow-cooker                                                                           ~/work/linkerd2
NAME                                                         APEX                                                         LEAF          WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc     500m   100.00%   0.4rps           1ms           2ms           2ms
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   failing-svc     500m     0.00%   0.6rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 main [📦📝🤷‍] via 🐼 v1.16.6 via  took 2s
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp --from deploy/slow-cooker-1                                                                         ~/work/linkerd2
NAME                                                         APEX                                                         LEAF          WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc     500m   100.00%   0.5rps           1ms           2ms           2ms
backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local   failing-svc     500m     0.00%   0.5rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 main [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/prometheus -n linkerd-viz                                                   ~/work/linkerd2
StatSummary API error: service only supported as a target on 'from' queries, or as a destination on 'to' queries%


# With no `sp.dstOverrides`

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via  took 10s
➜ k -n linkerd-trafficsplit-test-sp delete sp backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local                                     ~/work/linkerd2
serviceprofile.linkerd.io "backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local" deleted

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp                                                                  ~/work/linkerd2
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc        -   100.00%   1.2rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp --from deploy/slow-cooker-1 --from-namespace linkerd-trafficsplit-test-sp
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc        -   100.00%   0.6rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp --from deploy/slow-cooker --from-namespace linkerd-trafficsplit-test-sp
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc        -   100.00%   0.7rps           1ms           2ms           2ms

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat deploy/slow-cooker -n linkerd-trafficsplit-test-sp --to svc/backend-svc                                          ~/work/linkerd2
No traffic found.

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜                                                                                                                                            ~/work/linkerd2


```

Note: _This means that we need documenation changes to
let the user know that the `viz stat` on a service are client
side metrics and would be missing metrics from unmeshed
clients._


Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

adleong commented:
I think this is definitely the right direction.

Rather than special casing this for when the service has a ServiceProfile and a non-empty list of dstOverrides, I think we can make this always be the behavior for services.  Taking a step back, a ServiceProfile is just a mechanism for attaching extra information to a Service -- in this case for the purposes of client-side routing.  So we can think of a Service as always having a list of Dsts: either the dstOverrides from the associated ServiceProfile or, if that doesn't exist, just a single dst which is the service itself.

So I think when doing a `linkerd viz stat svc/foo` type of query, we should be consistent and output data like:

```
NAME          DST           WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-svc   backend-svc     500m   100.00%   0.5rps           1ms           2ms           2ms
backend-svc   failing-svc     500m     0.00%   0.5rps           1ms           2ms           2ms
```

More concretely, this means that for any service query, we'll look at the outbound metrics where the authority matches the service we're querying for.  We can then look at the `dst_service` values to find all the concrete dsts (in the case of a non-split service, there will be just one).

One thing to notice is that we don't have to introduce ServiceProfiles as a stat resource type; we can do everything in terms of services (though we probably have to look up the associated ServiceProfile in order to find the weights).

Taking things a step further, we can start to do some cleanup here by getting rid of all references to TrafficSplits in the main code path.  For now we can retain backwards compatibility by translating any query for `linkerd viz stat ts/foo` into `linkerd viz stat svc/foo-apex` where `foo-apex` is the apex service of the foo TrafficSplit.  This should yield the exact same behavior as today.  This is a good opportunity to do this since you'll be touching all of that logic anyway to accommodate services.

Let me know if this makes sense.

adleong commented on viz/metrics-api/stat_summary.go:
```diff
@@ -113,6 +110,8 @@ func (s *grpcServer) StatSummary(ctx context.Context, req *pb.StatSummaryRequest
 				resultChan <- s.nonK8sResourceQuery(ctx, statReq)
 			} else if isTrafficSplitQuery(statReq.GetSelector().GetResource().GetType()) {
 				resultChan <- s.trafficSplitResourceQuery(ctx, statReq)
+			} else if s.isServiceProfileQuery(statReq) {
```
Following the logic from my comment, I think that would mean combining this and the previous branch into one:

```
else if s.isServiceQuery(statReq) {
    resultChan <- s.serviceResourceQuery(ctx, statReq)
```

Pothulapati commented:
@adleong Yep, That makes sense. I updated the PR and description to reflect your suggestion i.e to also support `viz stat svc` using authority labels. 

This raises the following questions:

- Are  `viz stat <> --from svc/xyz` and `viz stat svc/xyz --to <>` supported? because those paths means that the service object is the client, which is not possible to query using authority as the main label. In the PR currently, Both of these cases return a error as any request with a `svc` object irrespective of `--from`, `--to` reaches `serviceResosurceQuery` code path. 
- Though, For the former in above, We can make it go into `k8sresourceQuery` path instead, where the same query is performed using `dst_service` with no `authority`, which is how it currently works.
- Currently, The output for svc with `sp.dstOverrides` is a bit confusing because the  `apex` is an authority while `leaf` is not, and is a plain service name with no namespace. This is because we render it directly from `dst_service`. Should we update it to be authority by attaching the namespace and cluster-domain. I'm more biased towards authorities because that is what the users uses when specifying the leafs in a sericeprofile's `dstOverrides`

Thanks

adleong commented:
* `viz stat <> --from svc/xyz`: I don't think this query is meaningful since it asks about traffic from a service and services cannot send traffic.  This already is considered invalid today

```console
> linkerd viz stat -n default deploy/authors --from svc/webapp
StatSummary API response error: service only supported as a target on 'from' queries, or as a destination on 'to' queries
```

* `viz stat svc/xyz --to <>`: This query IS meaninful because it asks about traffic sent to a service with an additional filter narrowing down the workload it's sent to.  This works today and I think should continue to work.  The `--to` flag is translated into a label constraint like `dst_deployment=foo` for example which can be combined with the rest of the query for `stat svc/xyz`.
* "Apex" and "Leaf" are SMI terminology which we don't need to use here.  I'd suggest simply a "Name" column for the service name (i.e. what SMI would call the Apex) and a "Dst" column for the concrete service (i.e. the dst_service).  I don't think we need to format either of these as authorities.

adleong commented:
I think this can still be significantly streamlined and simplified.  Take a look at https://github.com/linkerd/linkerd2/compare/tarun/cli-stat-sp...alex/service-stat-unification?expand=1 to see what I had in mind.

In particular:
* The StatSummary API no longer supports traffic split queries.  Instead, the CLI will convert a traffic split query into a service query for the apex.
* We don't have separate codepaths for services and service profiles.  Service is the resource that we operate on (which translates into an authority query grouped by dst_service).  The service stats are augmented by a service profile lookup to add the weights.
* We drop the "apex" concept altogether since the apex is the same as the service name.  We rename "leaf" to "dst".

The result is a single non-branching codepath for services in the StatSummary implementation.

Some sample output:

```console 
# A service with multiple dstOverrides defined in a service profile
> linkerd viz stat svc/voting-svc -n emojivoto
NAME         APEX         LEAF         WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
voting-svc   voting-svc   emoji-svc      100m   100.00%   0.0rps           0ms           0ms           0ms   0
voting-svc   voting-svc   voting-svc     900m    85.96%   0.9rps           1ms           3ms           4ms   0

# A regular service with no service profile
> linkerd viz stat -n emojivoto svc/web-svc
NAME      DST       WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
web-svc   web-svc             86.67%   2.0rps           3ms          10ms          18ms   0

# A service with a TrafficSplit
> linkerd viz stat -n emojivoto svc/emoji-svc
NAME        DST         WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emoji-svc   emoji-svc            100.00%   1.5rps           1ms           1ms           1ms   0
emoji-svc   web-svc                0.00%   0.2rps           0ms           0ms           0ms   0

# A TrafficSplit query (the CLI translates this into a service query)
> linkerd viz stat -n emojivoto ts/emoji-split
NAME        DST         WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
emoji-svc   emoji-svc            100.00%   1.5rps           1ms           1ms           1ms   0
emoji-svc   web-svc                0.00%   0.3rps           0ms           0ms           0ms   0
```

One small regression here is that we don't show weights for splits which came from a traffic split.  This is because the weighted dsts are not copied onto the service profile and will be fixed by the smi-controller.

Pothulapati commented:
@adleong 

I've looked at the changes, and also incorporated them into this PR. I have the following updates.

- Good suggestion that we can just merge `Services` and `ServiceProfiles` path to use the same metric functions as they are pretty similar. I updated the PR to do that. Now, All of them go into the same path, and use the same queries. Except we attach weights when there is a `ServiceProfile` available.

- Apart from the changes that you mentioned, I've also updated `stat_summary` to include the following changes. Feel free to mention if you have any concerns on these.
   - [Use `dst_namespace` label alsoo in `group_by` ](https://github.com/linkerd/linkerd2/blob/tarun/cli-stat-sp/viz/metrics-api/stat_summary.go#L463) to not add metrics with the same `dst_namespace` but on a different namespace.
   - [Add support for TCP metrics](https://github.com/linkerd/linkerd2/blob/tarun/cli-stat-sp/viz/metrics-api/stat_summary.go#L521) by using the same labels, visible with -owide` in the CLI.
   - [Update `ServiceResourceQuery` to skip setting `TrafficSplitStats` when weight is empty](https://github.com/linkerd/linkerd2/blob/tarun/cli-stat-sp/viz/metrics-api/stat_summary.go#L315). This means that, CLi won't add additional `weight` and `DST` columns when there are no weights. (i.e Service without ServiceProfile case).

Now, w.r.t `TrafficSplits`, I understand by having a conversion logic in `stat.go` from TS -> SP, we can simplify the logic but this means that we are compromising on the UX w.r.t no weights. This could be even more confusing with the added usage of `DST` instead of apex and leaf for TrafficSplit users.
My suggestion would be to instead keep the current TrafficSplits UX and logic (in stats_summary) as is, and remove it completely once we start pushing the `SMI` extension more. this means that there is no degraded experience on UX, and users would move into the SP UX once they start using the SMI extension with TS. WDYT?

Pothulapati commented:
Updated the PR to include support for `TrafficSplit` resources as is, by restoring some of the removed logic.

adleong commented:
Trying this out on a regular service that has no ServiceProfile and no TrafficSplit:

```console
linkerd viz stat deploy -n emojivoto
NAME       MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
emoji         1/1   100.00%   2.3rps           1ms           1ms           1ms          3
vote-bot      1/1   100.00%   0.3rps           1ms           1ms           1ms          1
voting        1/1    87.01%   1.3rps           1ms           1ms           1ms          3
web           1/1    92.75%   2.3rps           2ms           3ms           4ms          3
> linkerd viz stat svc -n emojivoto
No traffic found.
```

I'm not seeing any service stats

adleong reviewed (commented):
I did some more testing with this and noticed that when testing with emojivoto, I get stats for the emoji-svc and voting-svc, but not for the web-svc:

```
bin/linkerd viz stat svc -n emojivoto
NAME         MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
emoji-svc         -   100.00%   2.0rps           1ms           1ms           1ms          1
voting-svc        -    90.00%   1.0rps           1ms           2ms           3ms          1
```

Digging in, it looks like the problem is this query: 

```
authority=~\"^web-svc.emojivoto.svc.cluster.local.+\"
```

vote-bot uses the default http port (80) when calling web and thus does not specify the port in the authority.  This means the authority looks like `authority="web-svc.emojivoto.svc.cluster.local"` and does not match the above regex.  I think the regex needs to be relaxed to accommodate that the authority does not necessarily contain a port.

adleong reviewed (approved):
Works great!  Looking forward to cleaning this up further after 2.11

alpeb reviewed (commented):
Sorry I didn't follow the earlier discussion closely. So we're not yet surfacing weights in `linkerd stat` queries targeted at services that have dst_overrides? That's what I'm observing from my testing...

Pothulapati commented:
> So we're not yet surfacing weights in linkerd stat queries targeted at services that have dst_overrides? That's what I'm observing from my testing...

@alpeb That does seem to be the case from my testing. 🤔 

```bash
on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
backend-svc        -   100.00%   1.0rps           1ms           2ms           2ms          1

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
backend-svc        -   100.00%   1.0rps           1ms           2ms           2ms          1

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ k apply -f ./test/integration/trafficsplit/testdata/sp/updated-traffic-split-leaf-weights.yaml -n linkerd-trafficsplit-test-sp
serviceprofile.linkerd.io/backend-svc.linkerd-trafficsplit-test-sp.svc.cluster.local created

on ⛵ kind-kind  linkerd2 on 🌱 taru [📦📝🤷‍] via 🐼 v1.16.6 via 
➜ ./bin/go-run cli viz stat svc/backend-svc -n linkerd-trafficsplit-test-sp                                                        ~/work/linkerd2
NAME          DST           WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
backend-svc   backend-svc     500m   100.00%   0.5rps           1ms           2ms           2ms          1
backend-svc   failing-svc     500m     0.00%   0.5rps           1ms           2ms           2ms          0
```

Pothulapati commented on test/integration/stat/stat_test.go:
No specific difference/purpose. can probably be removed! WDYT?

alpeb replied:
Yeah, in particular I didn't get the intent of that duped deployment in `application.yaml`

alpeb commented:
> @alpeb That does seem to be the case from my testing.

```console
$ bin/linkerd viz stat ts
NAME                    APEX          LEAF          WEIGHT   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
backend-traffic-split   backend-svc   backend-svc      500   100.00%   0.6rps           1ms           1ms           1ms
backend-traffic-split   backend-svc   failing-svc      500     0.00%   0.4rps           1ms           1ms           1ms

$ bin/linkerd viz stat -n default svc/backend-svc
NAME          MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN
backend-svc        -     0.00%   0.5rps           1ms           1ms           1ms          1
```
It's not for me, the only difference is that I deployed that in the default namespace. I'll keep on testing...

alpeb reviewed (approved):
Ok I wasn't installing the ServiceProfile, thinking we already had synthetic ServiceProfiles with the weights being returned by Destination using TrafficSplit data, but of course that isn't surfaced into the objects that the metrics-api queries. So this LGTM :+1:
